Add in support for the version tag.

clalancette opened the pull request:
Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

Just like in [urdfdom](https://github.com/ros/urdfdom/pull/133), add support for the version tag in the URDF XML.  Note that I ended up adding the `version` tag by hand to all of the round-trip tests so that they pass; there are other tests that test the lack of a `version` tag, so I think this is fine.  @sloretz @scpeters FYI.

clalancette committed:
Add in support for the version tag.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

scpeters commented on test/test_urdf.py:
```diff
@@ -181,6 +181,30 @@ def test_link_multiple_collision(self):
+        xml = '''<?xml version="1.0"?>
+<robot name="test" version="foo">
+</robot>'''
+        #self.assertRaises(xmlr.core.ParseError, self.parse, xml)
```
can this commented line be deleted?

clalancette replied:
Yeah, definitely, just a leftover from testing.  Done in 9f78b7f

scpeters commented:
this also looks good; just one more minor comment

scpeters commented:
The concerns about parsing floats in https://github.com/ros/urdfdom/pull/133#issuecomment-557274975 are relevant here too

scpeters commented:
I think we need to update the version parsing here too

clalancette commented:
> I think we need to update the version parsing here too

Yeah, definitely,  I just ran out of time yesterday.  I'll get to this early next week.

clalancette committed:
Switch to using a string for the version number.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
> Yeah, definitely, I just ran out of time yesterday. I'll get to this early next week.

OK, late this week, but I did finally get it done.  I switched this to a string now, and updated the tests to match the C++ ones in 9f78b7f.  This is ready for another review.

scpeters commented on test/test_urdf.py:
```diff
@@ -181,6 +181,77 @@ def test_link_multiple_collision(self):
+
+    def test_version_attribute_no_major_number(self):
+        xml = '''<?xml version="1.0"?>
+<robot name="test" version="1.">
```
for `no_major_number`, should this be `version=".0"`?

clalancette replied:
Yeah, definitely.  That was a mistake, and it actually showed a bug.  Will fix.

scpeters commented on test/test_urdf.py:
```diff
@@ -181,6 +181,77 @@ def test_link_multiple_collision(self):
+
+    def test_version_attribute_negative_minor_number(self):
+        xml = '''<?xml version="1.0"?>
+<robot name="test" version="-1.0">
```
`1.-0`?

clalancette replied:
Sigh, yes.

sloretz commented on src/urdf_parser_py/urdf.py:
```diff
@@ -533,6 +533,20 @@ def get_root(self):
+
+        split = self.version.split(".")
+        if len(split) != 2:
+            raise Exception("The version attribute should be in the form 'x.y'")
```
Recommend a custom exception, but `ValueError` would fit too. Same below.

clalancette replied:
Yeah, I went back and forth on making a custom exception.  I think I'll go with "ValueError", as that seems to make the most sense to me.

sloretz commented on src/urdf_parser_py/urdf.py:
```diff
@@ -533,6 +533,20 @@ def get_root(self):
         assert root is not None, "No roots detected, invalid URDF."
         return root
 
+    def post_read_xml(self):
```
What calls this function? I can't find any uses of it.

clalancette replied:
It's magic :).  The `Robot` class is a sub-class of `xmlr.Object`, which defines an empty version of `post_read_xml`.  In turn, this method is called directly after reading in the XML (see [here](https://github.com/ros/urdf_parser_py/blob/b213e6e5846800b4873be349d605b906e78d681e/src/urdf_parser_py/xml_reflection/core.py#L590)).  So I just overrode it in the `Robot` sub-class.

sloretz commented on src/urdf_parser_py/urdf.py:
```diff
@@ -547,7 +561,8 @@ def from_parameter_server(cls, key='robot_description'):
 
 xmlr.reflect(Robot, tag='robot', params=[
-    xmlr.Attribute('name', str, False),  # Is 'name' a required attribute?
+    xmlr.Attribute('name', str),
```
Is making the `name` attribute required an intentional change?

clalancette replied:
I did intentionally change it.  The xsd specifies it as required: https://github.com/ros/urdfdom/blob/3ffc4eebb22e24bef503849eb43719a7c148e999/xsd/urdf.xsd#L318 .

In normal circumstances, I wouldn't attempt to change this in an already released distribution like melodic.  But I think this case is somewhat special, because the C++ URDF parser *already* enforces having a name attribute at https://github.com/ros/urdfdom/blob/3ffc4eebb22e24bef503849eb43719a7c148e999/urdf_parser/src/model.cpp#L113 .  Thus, it wouldn't be possible to get too far in the ROS ecosystem without a name for the robot.  So I think it is justified to change it here.  Thoughts?

clalancette committed:
Fixes from review.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette commented:
Thanks for the reviews!  I've made fixes corresponding to the reviews in e832a00.  @sloretz @scpeters ready for another round.

clalancette commented:
Thanks!  Merging.
